pallets/identity/src/lib.rs

```diff
@@ -1651,28 +1646,48 @@ impl<T: Trait> Module<T> {
         proof: InvestorZKProofData,
         expiry: Option<T::Moment>,
     ) -> DispatchResult {
+        let (scope, scope_id, cdd_id) =
+            if let Claim::InvestorUniqueness(scope, scope_id, cdd_id) = &claim {
+                (scope, scope_id, cdd_id)
+            } else {
+                return Err(Error::<T>::ClaimVariantNotAllowed.into());
+            };
         // Only owner of the identity can add that confidential claim.
         ensure!(
             issuer == target,
             Error::<T>::ConfidentialScopeClaimNotAllowed
         );
-
-        if let Claim::InvestorUniqueness(.., cdd_id) = &claim {
-            // Verify the owner of that CDD_ID.
-            ensure!(
-                Self::base_fetch_cdd(target, T::Moment::zero(), Some(*cdd_id)).is_some(),
-                Error::<T>::ConfidentialScopeClaimNotAllowed
-            );
-        }
+        // Verify the owner of that CDD_ID.
+        ensure!(
+            Self::base_fetch_cdd(target, T::Moment::zero(), Some(*cdd_id)).is_some(),
+            Error::<T>::ConfidentialScopeClaimNotAllowed
+        );
         // Verify the confidential claim.
         ensure!(
             ValidProofOfInvestor::evaluate_claim(&claim, &target, &proof),
             Error::<T>::InvalidScopeClaim
         );
 
-        if let Claim::InvestorUniqueness(Scope::Ticker(scope), scope_id, _) = &claim {
+        if let Scope::Ticker(ticker) = &scope {
+            // If `target` already has different InvestorUniqueness claims, clean up the old ScopeId data.
+            let old_scope_ids = Self::fetch_base_claims(target, ClaimType::InvestorUniqueness)
+                .filter_map(move |id_claim| {
```

Comment:
This is fine, but I'd probably write this using `match` + a guard myself in the current language.